packages/smart-contracts/src/contracts/BatchNoConversionPayments.sol

```diff
@@ -24,284 +25,462 @@ contract BatchNoConversionPayments is Ownable {
+      amountAndFee += requestDetails[i].requestAmount + requestDetails[i].feeAmount;
+      batchFeeAmount += requestDetails[i].requestAmount;
+    }
+    batchFeeAmount = (batchFeeAmount * batchFee) / feeDenominator;
+
+    // batchFeeToPay and batchFeeAmountUSD are updated if needed
+    (batchFeeAmount, batchFeeAmountUSD) = calculateBatchFeeToPay(
+      batchFeeAmount,
+      requestDetails[0].path[0],
+      batchFeeAmountUSD,
+      pathsToUSD
     );
 
-    // amount is used to get the total amount and fee, and then used as batch fee amount
-    uint256 amount = 0;
-    for (uint256 i = 0; i < _recipients.length; i++) {
-      amount += _amounts[i] + _feeAmounts[i];
-    }
+    IERC20 requestedToken = IERC20(requestDetails[0].path[0]);
 
-    // Transfer the amount and fee from the payer to the batch contract
-    IERC20 requestedToken = IERC20(_tokenAddress);
-    require(
-      requestedToken.allowance(msg.sender, address(this)) >= amount,
-      'Insufficient allowance for batch to pay'
+    contractAllowanceApprovalTransfer(
+      requestedToken,
+      amountAndFee,
+      batchFeeAmount,
+      address(paymentErc20Proxy)
     );
-    require(requestedToken.balanceOf(msg.sender) >= amount, 'not enough funds');
+
+    // Payer pays batch fee amount
     require(
-      safeTransferFrom(_tokenAddress, address(this), amount),
-      'payment transferFrom() failed'
+      safeTransferFrom(requestDetails[0].path[0], feeAddress, batchFeeAmount),
+      'Batch fee transferFrom() failed'
     );
 
-    // Batch contract approve Erc20FeeProxy to spend the token
-    if (requestedToken.allowance(address(this), address(paymentErc20Proxy)) < amount) {
-      approvePaymentProxyToSpend(address(requestedToken), address(paymentErc20Proxy));
-    }
-
     // Batch contract pays the requests using Erc20FeeProxy
-    for (uint256 i = 0; i < _recipients.length; i++) {
-      // amount is updated to become the sum of amounts, to calculate batch fee amount
-      amount -= _feeAmounts[i];
+    for (uint256 i = 0; i < requestDetails.length; i++) {
+      RequestDetail memory rD = requestDetails[i];
       paymentErc20Proxy.transferFromWithReferenceAndFee(
-        _tokenAddress,
-        _recipients[i],
-        _amounts[i],
-        _paymentReferences[i],
-        _feeAmounts[i],
-        _feeAddress
+        rD.path[0],
+        rD.recipient,
+        rD.requestAmount,
+        rD.paymentReference,
+        rD.feeAmount,
+        feeAddress
       );
     }
 
-    // amount is updated into batch fee amount
-    amount = (amount * batchFee) / tenThousand;
-    // Check if the payer has enough funds to pay batch fee
-    require(requestedToken.balanceOf(msg.sender) >= amount, 'not enough funds for the batch fee');
-
-    // Payer pays batch fee amount
-    require(
-      safeTransferFrom(_tokenAddress, _feeAddress, amount),
-      'batch fee transferFrom() failed'
-    );
+    return batchFeeAmountUSD;
```

Comment:
Shouldn't you always check that the contract balance is zero before returning?

Reply:
Not in the past, but now yes, good catch thx.
(before, we were paying the exact amount of fee, now, it can occur that it decreases... I will check other ERC20 function.)